#### 패션 MNIST

In [1]:
from tensorflow import keras

(train_input, train_target), (test_input, test_target) = \
    keras.datasets.fashion_mnist.load_data()

In [4]:
# train 과 valid data 분리
from sklearn.model_selection import train_test_split

train_scaled = (train_input / 255.0).reshape(-1, 28*28)     # 정규화 + 2차

train_scaled, val_scaled, train_target, val_target = \
    train_test_split(
        train_scaled,
        train_target,
        test_size=0.2,
        random_state=42
    )


#### Layer 를 추가하는 방법 - 1

In [ ]:
# Layer 만들기
dense1 = keras.layers.Dense(100, activation='sigmoid', input_shape=(784,))  # 은닉층
dense2 = keras.layers.Dense(10, activation='softmax')   # 출력층. 다중 출력이면 거의 softmax

c:\Users\tjoeun\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [6]:
# 전체 모델에 층을 추가
model = keras.Sequential([dense1, dense2])

In [7]:
# Summary
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 100)            │        78,500 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │         1,010 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 79,510 (310.59 KB)

 Trainable params: 79,510 (310.59 KB)

 Non-trainable params: 0 (0.00 B)

----
#### Layer 를 추가하는 방법 - 2

In [8]:
model = keras.Sequential(
    [
        keras.layers.Dense(100, activation='sigmoid', input_shape = (784,), name = 'hidden'),
        keras.layers.Dense(10, activation='softmax', name='output')
    ],
    name='패션 MNIST 모델'
)

c:\Users\tjoeun\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [9]:
model.summary()

Model: "패션 MNIST 모델"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ hidden (Dense)                  │ (None, 100)            │        78,500 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ output (Dense)                  │ (None, 10)             │         1,010 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 79,510 (310.59 KB)

 Trainable params: 79,510 (310.59 KB)

 Non-trainable params: 0 (0.00 B)

----
#### Layer 를 추가하는 방법 - 3

In [13]:
model = keras.Sequential()
model.add(keras.layers.Dense(100, activation='sigmoid', input_shape = (784, ), name='hidden'))
model.add(keras.layers.Dense(10, activation='softmax', name='output'))

c:\Users\tjoeun\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [14]:
model.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ hidden (Dense)                  │ (None, 100)            │        78,500 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ output (Dense)                  │ (None, 10)             │         1,010 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 79,510 (310.59 KB)

 Trainable params: 79,510 (310.59 KB)

 Non-trainable params: 0 (0.00 B)

> 제일 많이 사용하는 방법. 파이썬 코드가 중간에 삽입이 가능하기 때문. (ex_if문으로 분기 줄 수 있음)

In [15]:
model.compile(
    loss = 'sparse_categorical_crossentropy',
    metrics = ['accuracy']
)
model.fit(train_scaled, train_target, epochs=5)

Epoch 1/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8078 - loss: 0.5692
Epoch 2/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8519 - loss: 0.4125
Epoch 3/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8633 - loss: 0.3765
Epoch 4/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8717 - loss: 0.3537
Epoch 5/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8777 - loss: 0.3356


-----
#### 입력층을 이미지 그대로 사용하기
(train_scaled = (train_input / 255.0).reshape(-1, 28*28) 에서 reshape 를 안 거치고 만들기)

In [16]:
(train_input, train_target), (test_input, test_target) = \
    keras.datasets.fashion_mnist.load_data()

train_scaled = train_input / 255.0

train_scaled, val_scaled, train_target, val_target = \
    train_test_split(
        train_scaled,
        train_target,
        test_size=0.2,
        random_state=42
    )

In [17]:
train_input.shape

(60000, 28, 28)

In [57]:
model = keras.Sequential()

model.add(keras.layers.Flatten(input_shape = (28, 28)))   # flatten : 펴주는 함수
model.add(keras.layers.Dense(100, activation='relu'))
model.add(keras.layers.Dense(10, activation='softmax'))     # 둘 중 하나면 시그모이드

In [19]:
model.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten (Flatten)               │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 100)            │        78,500 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 10)             │         1,010 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 79,510 (310.59 KB)

 Trainable params: 79,510 (310.59 KB)

 Non-trainable params: 0 (0.00 B)

In [58]:
# 손실함수
import numpy as np

model.compile(
    optimizer = 'adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)


In [59]:
# 학습하기
model.fit(np.array(train_scaled), np.array(train_target), epochs = 16)     # 현재 pandas 데이터인데 numpy 로 바꿔서 훈련해야 함

Epoch 1/16
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8182 - loss: 0.5220
Epoch 2/16
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8587 - loss: 0.3975
Epoch 3/16
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8712 - loss: 0.3549
Epoch 4/16
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8829 - loss: 0.3254
Epoch 5/16
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8874 - loss: 0.3081
Epoch 6/16
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8925 - loss: 0.2914
Epoch 7/16
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8978 - loss: 0.2779
Epoch 8/16
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.9019 - loss: 0.2665
Epoch 9/16
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.9052 - loss: 0.2560
Epoch 10/16
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.9081 - loss: 0.2469
Epoch 11/16
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.9112 - loss: 0.2399
Epoch 12/16
1500/1500 ━━━━━━━━

In [60]:
# 검증 데이터로 확인
print(model.evaluate(val_scaled, val_target)[1])

375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.8863 - loss: 0.3337
0.8863333463668823


In [61]:
# 테스트로 확인 . 테스트는 / 255 를 안 함. 정규화 해줘야 함.
test_scaled = test_input / 255.0
print(model.evaluate(train_scaled, train_target)[1])
print(model.evaluate(test_scaled, test_target)[1])

1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.9279 - loss: 0.1906
0.9278958439826965
313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8807 - loss: 0.3632
0.8806999921798706


val / train / test
좋은 조건 : 1. train - test 차이 5% 미만    2. test 가 제일 높은 거

epochs = 5
0.8767499923706055 / 0.8959375023841858 / 0.869700014591217

epochs = 10
0.8870833516120911 / 0.9177291393280029 / 0.8810999989509583

epochs = 15
0.8855833411216736 / 0.924791693687439 / 0.8805000185966492

epochs = 16
0.8863333463668823 / 0.9278958439826965 / 0.8806999921798706        => 최적

epochs = 17
0.8859999775886536 / 0.9256666898727417 / 0.8799999952316284

epochs = 18
0.8913333415985107 / 0.9367291927337646 / 0.8858000040054321

epochs = 20
0.8921666741371155 / 0.9378125071525574 / 0.885699987411499

epochs = 30
0.8842499852180481 / 0.9417083263397217 / 0.8780999779701233  => 5% 이상 차이 = X

 - 은닉층이 여러 개 생길 수록 느려지나, 정확도는 올라감